component/main.jsonnet

```
@@ -8,44 +8,71 @@ local params = inv.parameters.crossplane;
 local on_openshift4 = inv.parameters.facts.distribution == 'openshift4';
 local has_service_account(provider) = std.count(std.objectFields(params.serviceAccounts), provider) > 0;
 local has_any_service = std.length(std.filter(function(x) std.objectHas(params.serviceAccounts, x), std.objectFields(params.providers))) > 0;
+local missing_controller(provider_name) = std.count(std.objectFields(params.controllerConfigs), provider_name) == 0;
+local name_from_resource(resource) = com.getValueOrDefault(com.getValueOrDefault(resource, 'metadata', ''), 'name', '');
```

I don't think this works as you want, since `com.getValueOrDefault()` expects the first argument to be an object. You need to return an empty object as the default when using `com.getValueOrDefault()` to extract an object:

```suggestion
local name_from_resource(resource) = com.getValueOrDefault(com.getValueOrDefault(resource, 'metadata', {}), 'name', '');
```

---

I think this works as I expected it as I had done various tests to make sure it does what I want. I always expect that the name is there because otherwise the function `com.generateResources(params.controllerConfigs, crossplane.ControllerConfig)` would get me an error in case it cannot parse the object, am I right?

---

yes the resulting objects of `com.generateResources()` will always have `.metadata.name`, so you don't actually need the function at all and can simply use `merge_service_account_from_resource(controller_config.metadata.name)` instead of calling `name_from_resource()`

---

that was a stupid miss observation from my side. Thanks for suggestion, I checked and it works!